lib/PipelineC/add_tracing.py

```diff
@@ -0,0 +1,205 @@
+}
+
+# Template for specifying length hints to the C++ backend
+LH_DECL = 'template<> inline constexpr int LengthHint<"{}", {}> = {};\n'
+
+# Template that will be used to implement the `isRPType` function that will
+# tell if a type is part of PipelineC at compile time
+TYPE_DECL = "using RPTypes = std::tuple<{}>;\n"
+
+
+@dataclass
+class Argument:
+    # Name of the argument (e.g. "foo")
+    name: str
+    # Declaration of the argument (with type, e.g. "int *foo[]")
+    decl: str
+
+
+@dataclass
+class Function:
+    # Name of the function
+    name: str
+    # List of arguments
+    arguments: List[Argument]
+    # Return type
+    return_type: str
+
+
+class Visitor(NodeVisitor):
+    def __init__(self):
+        super().__init__()
+        self.c_gen = CGenerator()
+        self.functions: List[Function] = []
+        self.types: List[str] = []
+
+    def visit_FuncDecl(self, node: FuncDecl):  # noqa: N802
+        name = node.type.declname if isinstance(node.type, TypeDecl) else node.type.type.declname
+        return_type = self.c_gen.visit(node.type)
+        arguments = []
+        if node.args is not None:
+            arguments = [self.parse_argument(arg) for arg in node.args.params]
+        self.functions.append(Function(name, arguments, return_type))
+
+    def visit_Typedef(self, node: Typedef):  # noqa: N802
+        if node.name not in {"bool", "uint8_t", "uint32_t", "uint64_t"}:
+            self.types.append(node.name)
+
+    def parse_argument(self, arg: str) -> Argument:
+        string: str = self.c_gen.visit(arg)
+        name_match = name_re.search(string.removesuffix("[]"))
+        assert name_match is not None
+        return Argument(name_match[0], string)
+
+
+def under_functions(text, functions: List[Function]):
+    new_text = text[:]
+    for function in functions:
+        target = re.search(function.name + r"\([^\)]*\)\s*{", new_text, re.DOTALL | re.MULTILINE)
+        if target:
+            new_text = new_text[: target.start()] + "_" + new_text[target.start() :]
+    return new_text
+
+
+def inject_trace_header(text: str):
+    lines = text.splitlines()
+    lines[lines.index("")] = '#include "TracingPrototypes.h"'
+    return "".join(f"{line}\n" for line in lines)
+
+
+def parse_headers(headers: List[Path]):
+    # Fixup definitions for types not present in C99
+    text = """
+typedef struct bool bool;
+typedef struct uint8_t uint8_t;
+typedef struct uint32_t uint32_t;
+typedef struct uint64_t uint64_t;
+"""
+    for header in headers:
+        text += header.read_text()
+
+    # Remove comments and pre-processor directives
+    text = comment_re.sub("", text)
+    text = "".join(ln if not ln.startswith("#") else "\n" for ln in text.splitlines(True))
+
+    parser = CParser()
+    visitor = Visitor()
+    visitor.visit(parser.parse(text, "input"))
+    return (visitor.functions, visitor.types)
+
+
+def generate(output: Path, impl_files: List[Path], functions: List[Function], types: List[str]):
```

Comment:
I'd like this script to follow the typical practices when generating files that are included in C++ as much as possible.

This means:

1. One `functions.inc` file that contains all the functions that are part of the C API:
    ```c++
    FUNCTION(rp_initialize)
    FUNCTION(rp_lol)
    ...
    ```

    Usage:

    ```c++
    #define FUNCTION(name) ...
    #include "functions.inc"
    #undef FUNCTION
    ```

    You might not like this, but this is a standard practice and in case we need to use it once more is configurable without touching the Python script.

2. One `types.inc` file listing all the types
    ```c++
    TYPE(rp_container)
    TYPE(rp_kind)
    ```

    Usage as above.
    You can use this to build the tuple.

    ```c++
    std::tuple<
    #define TYPE(name) name,
    #include "types.inc"
    #undef TYPE
    void>
    ```

    This could be done in C++ templates by using `function_traits` and concatenating + dedup'ing tuples. But I'm OK with generating this as of now.

Then there's a third exceptional file (`wrappers.inc`) that generates wrappers.
That's fine to be ad-hoc.

`PipelineC.cpp` should become something like:

```c++
// ...

// Note the the underscore in front of function names
static const char *_rp_buffer_data(const rp_buffer *buffer) {
  return buffer->data();
}

static void _rp_buffer_destroy(const rp_buffer *buffer) {
  delete buffer;
}

#include "wrappers.inc"
```

No need to `sed` `PipelineC.cpp`.
Users cannot accidentally forget about defining the `static` function with an `_` in front of it, since if the they define, e.g., `rp_initialize` instead of `_rp_initialize` they will get an error for `rp_initialize` being defined twice.

This also makes `TracingPrototypes.h` unnecessary.

Reply:
ACK
Do note that having `#include "wrappers.inc"` in PipelineC (vs. autogenerated `TracingPrototypes.h`) means that if we ever want to split `PipelineC.cpp` it'd be a bit more of an headache

Reply:
Implemented